### PR TITLE
Quaternion: Fix bug where operations are left unevaluated

### PR DESCRIPTION
**Pull request by maxencemayrand**
#### Brief description of what is fixed or changed 

There are several operations on quaternions which are left unevaluated
in some common situations. One of the most obviously problematic one
is the following, which is actually part of the tests on this module.

```python
>>> from sympy import Quaternion, symbols
>>> x, y, z, w = symbols('x y z w')
>>> q = Quaternion(x, y, z, w)
>>> q.normalize()
x + y*i + z*j + w*k/sqrt(w**2 + x**2 + y**2 + z**2)
>>> type(q.normalize())
<class 'sympy.core.mul.Mul'>
```

Here, `q.normalize()`, instead of returning a quaternion, returns an
unevaluated `Mul` object. The expected output would have been an
instance of `Quaternion` whose arguments are `x/sqrt(w**2 + x**2 +
y**2 + z**2)`, `y/sqrt(w**2 + x**2 + y**2 + z**2)`, etc.

The test for this is `assert q.normalize() == Quaternion(x, y, z, w) /
sqrt(w**2 + x**2 + y**2 + z**2)`. It passes, but both sides of the
equation are not quaternions, which is not what we would expect.

There are two reasons for this bug:

1. Division for quaternions is not implemented. For instance, `q / 2`
is left unevaluated.

2. Operations between a quaternion and an expression which is neither
a quaternion nor a complex number is left unevaluated. For example, `q*q.norm()**-1` is left unevaluated since `(q.norm()**-1).is_complex`
is `None`.

Both problems have been corrected.

#### Release Notes

* algebras
   * Implement division for quaternions and more flexible operations with other expressions.

### CODE REVIEW

File: sympy/algebras/quaternion.py [hunk 1]
```diff
@@ -244,14 +254,13 @@ def add(self, other):
...
                     return Quaternion(re(q2) + q1.a, im(q2) + q1.b, q1.c, q1.d)
-                else:
-                    # q2 is something strange, do not evaluate:
-                    return Add(q1, q2)
-            else:
+            elif q2.is_commutative:
                 return Quaternion(q1.a + q2, q1.b, q1.c, q1.d)
+            else:
+                # q2 is something strange, do not evaluate:
+                return Add(q1, q2)
```
Comment (jksuom) [inline sympy/algebras/quaternion.py]: Maybe ValueError should be raised. (Otherwise the result would be even more strange.)

Comment (maxencemayrand) [reply sympy/algebras/quaternion.py]: Good idea; I’ll do it. Thanks for reviewing.

Comment (maxencemayrand) [reply sympy/algebras/quaternion.py]: I did the changes. For some reason, this behaviour, i.e. returning unevaluated Mul objects when adding a non-commutative expression, was tested explicitly in the module. But I agree with you that it is much better to raise a ValueError, so I added that and I removed those tests.

File: sympy/algebras/tests/test_quaternion.py [hunk 2]
```diff
@@ -25,17 +25,12 @@ def test_quaternion_construction():
...
     b = symbols("b", real=True)
-    # This symbol is not complex:
-    c = symbols("c", commutative=False)
 
     q = Quaternion(x, y, z, w)
     assert a + q == Quaternion(x + re(a), y + im(a), z, w)
     assert 1 + q == Quaternion(1 + x, y, z, w)
     assert I + q == Quaternion(x, 1 + y, z, w)
     assert b + q == Quaternion(x + b, y, z, w)
-    assert c + q == Add(c, Quaternion(x, y, z, w), evaluate=False)
```
Comment (jksuom) [inline sympy/algebras/tests/test_quaternion.py]: It is not necessary to totally remove these tests. It should be possible to write something like `raises(ValueError, lambda: c + q)` instead.

Comment (maxencemayrand) [reply sympy/algebras/tests/test_quaternion.py]: Done. Thanks for reviewing again.

**Comment by jksuom** [comment]
Thanks, this looks good.

**Pull request merged**